File: Server/src/Match.cpp
#include "Match.hpp"

Match::Match(int IDMatch, NetClient* P1, NetClient* P2)
{
    this->IDMatch = IDMatch;
    this->players.push_back(P1);
    this->players.push_back(P2);
}

Match::~Match()
{
    for(auto player : this->players)
        delete player;

    this->players.clear();
}

void Match::LoadGame()
{
    // # Gera um seed.
    srand(time(nullptr));
    int start_turn = rand() % 2;

    // # Gera um seed.
    srand(time(nullptr));
    int my_tank_p1 = rand() % 2;
    int my_tank_p2 = my_tank_p1 == 1 ? 0 : 1;
    
    // # Monta pacote.
    auto base_pack = "10|" + std::to_string(IDMatch) + "|" + players[0]->Nickname + "|" + players[0]->Nickname + "|";

    // # Monta pacote para cada jogador.
    auto PID1_Packet = base_pack + std::to_string(my_tank_p1);
    auto PID2_Packet = base_pack + std::to_string(my_tank_p2);

    // # Envia o pacote de carregamento para cada jogador.
    socket.send(PID1_Packet.c_str(), PID1_Packet.size() + 1, players[0]->IP, players[0]->Port);
    socket.send(PID2_Packet.c_str(), PID2_Packet.size() + 1, players[1]->IP, players[1]->Port);
}

void Match::PushMessage(std::string strPackage)
{
    auto packet = Tools::Split(strPackage, '|');
    int netcode = stoi(packet[0]);

    // # Troca o turno dos jogadores.
    if(netcode == 4)
    {
        int playerID = stoi(packet[1]);
        playerID = playerID == 0 ? 1 : 0;
        strPackage = packet[0] + "|" + std::to_string(playerID);
    }

    // # Espera a confirmação de todos para iniciar a partida.
    if(netcode == 12)
    {
        int playerID = stoi(packet[1]);
        this->players[playerID]->isReady = true;

        if(this->players[0]->isReady && this->players[1]->isReady)
        {
            srand(time(nullptr));
            auto pack = "9|" + std::to_string(rand() % 2);
            this->SendPacket(pack);
            return;
        }
    }

    this->SendPacket(strPackage);

    // # Acabou a partida.
    if(netcode == 8)
        this->GameIsRunning = false;
}

void Match::SendPacket(std::string packet)
{
    for(auto player : this->players)
        socket.send(packet.c_str(), packet.size() + 1, player->IP, player->Port);
}

File: Server/match.mjs
export class Match
{
    ID = 0;
    GameIsRunning = true;
    Players = [];
    Current_turn = 0;
    ID_spectator_count = 0;
    minute = 4;
    second = 59;
    timerRef;

    constructor(initData)
    {
        this.ID = initData.IDMatch;
        this.Players.push(initData.PID1);
        this.Players.push(initData.PID2);
    }

    LoadGame()
    {
        // # Sorteia os tanks entre os jogadores.
        let sortRand = Math.floor(Math.random() * 5);

        for(let i = 0; i < sortRand; i++)
            this.Players.reverse();

        // # Carrega stats dos jogadores.
        this.Players[0].HP = 5;
        this.Players[0].position = "(79, 86)";
        this.Players[1].HP = 5;
        this.Players[1].position = "(1123, 78)";

        // # Monta pacote base.
        let base_pack = `10|${this.ID}|${this.Players[0].nickname}|${this.Players[1].nickname}|`;
        
        // # Monta pacote para cada jogador.
        let PID1_Packet = `${base_pack}0`;
        let PID2_Packet = `${base_pack}1`;

        // # Envia o pacote de carregamento para cada jogador.
        this.Players[0].con.send(PID1_Packet);
        this.Players[1].con.send(PID2_Packet);
    }

    AssignSpectator(con)
    {
        const idSpectator = ++this.ID_spectator_count;
        const timer = `${this.minute}-${this.second}`;
        this.Players.push({"idSpectator": idSpectator, "con": con});
        let current_game_state = `18|${idSpectator}|${this.Current_turn}|${timer}`;
        this.Players.forEach( e => { if(e["idSpectator"] == null) current_game_state += `|${e.nickname}-${e.HP}-${e.position}`; } );
        con.send(current_game_state);
    }

    RemoveSpectator(idSpectator)
    {
        this.Players = this.Players.filter(e => e.idSpectator != idSpectator);
    }

    PushMessage(msgPackage)
    {
        const chunks = msgPackage.toString().split("|");
        const netcode = Number(chunks[0]);

        // # Espera a confirmação de todos para iniciar a partida.
        if(netcode == 12)
        {
            let IDPlayer = Number(chunks[1]);

            this.Players[IDPlayer].IsReady = true;

            if(this.Players[0].IsReady && this.Players[1].IsReady)
            {
                this.Current_turn = Math.floor(Math.random() * 2);
                let pack = `9|${this.Current_turn}`;
                this.SendPacket(pack);
                // # Inicia a contagem de tempo.
                this.timerRef = setInterval(() => { this.TimerHandler(); }, 1000);
                return;
            }
        }

        // # Espera a confirmação de todos para fechar a partida.
        if(netcode == 16)
        {
            let IDPlayer = Number(chunks[1]);
            this.Players[IDPlayer].IsReady = true;

            if(this.Players[0].isCloseGame && this.Players[1].isCloseGame)
            {
                this.Players.forEach(e => { e.con.close(); e.con = null; });
                this.GameIsRunning = false;
            }
        }

        // # Muda o turno e sincroniza os dados.
        if(netcode == 5)
        {
            // # Armazena o turno do jogador atual.
            this.Current_turn = this.Current_turn == 1 ? 0 : 1;

            // # Obtém o estado da partida.
            const STATE_P1 = chunks[2].split('-');
            const STATE_P2 = chunks[3].split('-');

            // # Atualiza o estado do jogador 1.
            this.Players[0].HP = Number(STATE_P1[0]);
            this.Players[0].position = STATE_P1[1];

            // # Atualiza o estado do jogador 2.
            this.Players[1].HP = Number(STATE_P2[0]);
            this.Players[1].position = STATE_P2[1];

            const msgPackage = `5|${this.Current_turn}|${chunks[2]}|${chunks[3]}|${chunks[4]}|${this.minute}|${this.second}|${this.Players[this.Current_turn].nickname}`;
            this.SendPacket(msgPackage);
        }

        if(netcode == 8)
        {
            setTimeout(() => { this.GameIsRunning = false; this.minute = 0; this.second = 0; },1000);
        }

        // # Repassa os pacotes para os jogadores.
        this.SendPacket(msgPackage);
    }

    SendPacket(pack)
    {
        this.Players.forEach(e => { if(e.con) e.con.send(pack); });
    }
    
    TimerHandler()
    {
        this.second--;

        if(this.minute <= 0 && this.second <= 0)
        {
            clearInterval(this.timerRef);
            this.DefineWinner();
            return;
        }

        if(this.second <= 0)
        {
            this.minute--;
            this.second = 59;

            if(this.minute <= 0)
                this.minute = 0;
        }
    }

    DefineWinner()
    {
        const PID1 = this.Players[0];
        const PID2 = this.Players[1];
        let WinnerName = "Empate";
        
        if(PID1 && PID2)
        {   
            const PID1_HP = PID1.HP;
            const PID2_HP = PID2.HP;
            
            if(PID1_HP > PID2_HP)
                WinnerName = PID1.nickname;
            else if(PID1_HP < PID2_HP)
                WinnerName = PID2.nickname;
        }

        let pack = `8|${WinnerName}`;
        
        this.SendPacket(pack)
    }

    isGamaOver()
    {
        return !this.GameIsRunning || (this.minute <= 0 && this.second <= 0);
    }
}